Add remaining method and tests for dtype propagation

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #141495
* __->__ #140057
* #139945

Adds the remaining unimplemented ops as well as an assertion failure if someone adds a new op without a dtype rule.

We test all unique pointwise operators registered as lowerings which have an opinfo. There will be some follow ups for this to work well with both `codegen_upcast_to_fp32` as True and False. 


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

eellison committed:
Add remaining method and tests

[ghstack-poisoned]

blaine-rister commented:
Suprseded by https://github.com/pytorch/pytorch/pull/140864

blaine-rister commented:
Whoops, accidentally closed this

eellison committed:
Update on "Add remaining method and tests"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Add remaining method and tests"


Adds the remaining unimplemented ops as well as an assertion failure if someone adds a new op without a dtype rule.

We test all unique pointwise operators registered as lowerings which have an opinfo. There will be some follow ups for this to work well with both `codegen_upcast_to_fp32` as True and False. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

blaine-rister commented on torch/_inductor/dtype_propagation.py:
```diff
@@ -96,92 +107,235 @@ def __init__(self) -> None:
+
+    @staticmethod
+    def round_decimal(x: DTypeArg, y: DTypeArg) -> torch.dtype:
+        # TODO - dont see it anywhere..
```
Does this comment mean the op isn't actually used?

eellison replied:
Yea, i dont see a lowering for it and I dont know how it got here...

blaine-rister commented on torch/_inductor/codegen/common.py:
```diff
@@ -2024,7 +2025,24 @@ def do_cse(v):
+                            from torch._inductor.codegen.triton import triton_type
+
+                            if isinstance(output_dtype, (list, tuple)):
+                                output_dtype = output_dtype[idx]
```
Does this work because `do_cse` is called on each output in order by `pytree.tree_map`? It might be good to leave a comment explaining the logic since this is pretty tricky.

blaine-rister commented on torch/_inductor/codegen/common.py:
```diff
@@ -2024,7 +2025,24 @@ def do_cse(v):
                             dtype=output_dtype,
                         )
+
+                        nonlocal idx
```
```suggestion
                        nonlocal output_idx
```

Suggest a more informative name

blaine-rister commented on torch/_inductor/codegen/common.py:
```diff
@@ -2024,7 +2025,24 @@ def do_cse(v):
+                        nonlocal idx
+                        if (
+                            config.test_configs.runtime_triton_dtype_assert
+                            and V.graph.get_current_device_or_throw().type == "cuda"
```
Why only cuda? Does this break on non-GPU triton backends like CPU or MTIA?

eellison replied:
It can work for CPU-triton or mtia. I was trying to avoid cpu cpp codegen. I can update to be more generic.

blaine-rister replied:
Makes sense! I guess we could just exclude that specific backend then?

eellison committed:
Update on "Add remaining method and tests"


Adds the remaining unimplemented ops as well as an assertion failure if someone adds a new op without a dtype rule.

We test all unique pointwise operators registered as lowerings which have an opinfo. There will be some follow ups for this to work well with both `codegen_upcast_to_fp32` as True and False. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

blaine-rister reviewed (commented):
LGTM! Left a few nit comments about clarifying some of the trickier aspects.

blaine-rister commented:
nit: Should the PR title mention dtype propagation?

arui-meta commented:
LGTM! Thanks for working on this!

eellison committed:
Update on "Add remaining method and tests"


Adds the remaining unimplemented ops as well as an assertion failure if someone adds a new op without a dtype rule.

We test all unique pointwise operators registered as lowerings which have an opinfo. There will be some follow ups for this to work well with both `codegen_upcast_to_fp32` as True and False. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Add remaining method and tests for dtype propagation"


Adds the remaining unimplemented ops as well as an assertion failure if someone adds a new op without a dtype rule.

We test all unique pointwise operators registered as lowerings which have an opinfo. There will be some follow ups for this to work well with both `codegen_upcast_to_fp32` as True and False. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Add remaining method and tests for dtype propagation"


Adds the remaining unimplemented ops as well as an assertion failure if someone adds a new op without a dtype rule.

We test all unique pointwise operators registered as lowerings which have an opinfo. There will be some follow ups for this to work well with both `codegen_upcast_to_fp32` as True and False. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Add remaining method and tests for dtype propagation"


Adds the remaining unimplemented ops as well as an assertion failure if someone adds a new op without a dtype rule.

We test all unique pointwise operators registered as lowerings which have an opinfo. There will be some follow ups for this to work well with both `codegen_upcast_to_fp32` as True and False. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Add remaining method and tests for dtype propagation"


Adds the remaining unimplemented ops as well as an assertion failure if someone adds a new op without a dtype rule.

We test all unique pointwise operators registered as lowerings which have an opinfo. There will be some follow ups for this to work well with both `codegen_upcast_to_fp32` as True and False. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

eellison committed:
Update on "Add remaining method and tests for dtype propagation"


Adds the remaining unimplemented ops as well as an assertion failure if someone adds a new op without a dtype rule.

We test all unique pointwise operators registered as lowerings which have an opinfo. There will be some follow ups for this to work well with both `codegen_upcast_to_fp32` as True and False. 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]

eellison commented:
@pytorchbot merge
